"""
SMSå¾®æœåŠ¡ä¸»ç¨‹åº - å®Œå…¨ä¿®å¤ä¿¡å·å¤„ç†ç‰ˆæœ¬
"""
import asyncio
import signal
import socket
import sys
from concurrent import futures
from pathlib import Path
from typing import Optional
import grpc
from loguru import logger

from src.common.config import ConfigManager
from src.common.serial_detector import SerialDetector
from src.sms_service import sms_pb2_grpc
from src.sms_service.server import SMSService
from src.sms_service.sms_sender import SMSSender
from src.sms_service.consul_client import ConsulClient


class SMSMicroservice:
    """SMSå¾®æœåŠ¡ç®¡ç†å™¨"""

    def __init__(self, config_path: str = "config/sms.yaml"):
        self.config_path = Path(config_path)
        self.config: Optional[ConfigManager] = None
        self.serial_detector: Optional[SerialDetector] = None
        self.sms_sender: Optional[SMSSender] = None
        self.consul_client: Optional[ConsulClient] = None
        self.grpc_server: Optional[grpc.aio.Server] = None
        self._shutdown_event = asyncio.Event()
        self._shutting_down = False
        self._main_task: Optional[asyncio.Task] = None

    async def start(self) -> bool:
        """å¯åŠ¨å¾®æœåŠ¡"""
        try:
            logger.info("ğŸš€ å¯åŠ¨SMSå¾®æœåŠ¡...")

            # 1. åŠ è½½é…ç½®
            self.config = ConfigManager()
            if not await self.config.load_config(self.config_path):
                logger.error("âŒ é…ç½®åŠ è½½å¤±è´¥")
                return False

            cfg = self.config.get_config()

            # 2. é…ç½®æ—¥å¿—
            await self._setup_logging(cfg.log)

            # 3. æ‰“å°é…ç½®ä¿¡æ¯
            await self._print_config(cfg)

            # 4. æ£€æµ‹è°ƒåˆ¶è§£è°ƒå™¨
            logger.info("ğŸ“¡ æ£€æµ‹è°ƒåˆ¶è§£è°ƒå™¨...")
            self.serial_detector = SerialDetector(self.config)
            modems = await self.serial_detector.detect_modems()

            if not modems:
                logger.error("âŒ æœªæ£€æµ‹åˆ°å¯ç”¨çš„è°ƒåˆ¶è§£è°ƒå™¨")
                return False

            # é€‰æ‹©æœ€ä½³è°ƒåˆ¶è§£è°ƒå™¨
            best_modem = self.serial_detector.get_best_modem()
            if not best_modem:
                logger.error("âŒ æ— æ³•é€‰æ‹©è°ƒåˆ¶è§£è°ƒå™¨")
                return False

            logger.info(f"âœ… ä½¿ç”¨è°ƒåˆ¶è§£è°ƒå™¨: {best_modem.port} ({best_modem.manufacturer} {best_modem.model})")

            # 5. åˆå§‹åŒ–çŸ­ä¿¡å‘é€å™¨
            serial_config = cfg.serial
            self.sms_sender = SMSSender(
                port=best_modem.port,
                baudrate=serial_config.baudrate,
                timeout=serial_config.timeout
            )

            # è¿æ¥åˆ°è°ƒåˆ¶è§£è°ƒå™¨
            connected = await self.sms_sender.connect()
            if not connected:
                logger.error("âŒ è°ƒåˆ¶è§£è°ƒå™¨è¿æ¥å¤±è´¥")
                return False

            # 6. è§£æç›‘å¬åœ°å€
            host, port_str = cfg.server.listen_on.split(":")
            port = int(port_str)

            # å¦‚æœæ˜¯é€šé…ç¬¦åœ°å€ï¼Œè·å–æœ¬åœ°IP
            if host in ["0.0.0.0", "127.0.0.1", "[::]", "[::1]"]:
                host = socket.gethostbyname(socket.gethostname())

            # 7. æ³¨å†Œåˆ°Consul
            if cfg.consul.host and cfg.consul.host != "127.0.0.1:8500":
                logger.info(f"ğŸ”— è¿æ¥Consul: {cfg.consul.host}")

                server_data = {
                    "fields": {
                        "phone_number": {
                            "required": True,
                            "type": "string",
                            "pattern": r"^\+?[1-9]\d{1,14}$",
                            "description": "æ‰‹æœºå·ç ï¼ˆå›½é™…æ ¼å¼ï¼‰"
                        },
                        "content": {
                            "required": True,
                            "type": "string",
                            "maxLength": 1000,
                            "description": "çŸ­ä¿¡å†…å®¹"
                        },
                        "sender_id": {
                            "required": False,
                            "type": "string",
                            "description": "å‘é€è€…æ ‡è¯†"
                        },
                        "delivery_report": {
                            "required": False,
                            "type": "boolean",
                            "description": "æ˜¯å¦è¦æ±‚é€è¾¾æŠ¥å‘Š"
                        }
                    }
                }

                self.consul_client = ConsulClient(
                    host=cfg.consul.host,
                    token=cfg.consul.token,
                    scheme=cfg.consul.scheme
                )

                if await self.consul_client.register_service(
                    service_name=cfg.server.name,
                    address=host,
                    port=port,
                    service_desc="çŸ­ä¿¡å‘é€å¾®æœåŠ¡ï¼Œæ”¯æŒä¸­æ–‡çŸ­ä¿¡",
                    server_data=server_data,
                    meta={
                        "version": "1.0.0",
                        "modem_port": best_modem.port,
                        "modem_model": best_modem.model,
                        "signal": best_modem.signal_strength
                    }
                ):
                    logger.info("âœ… Consulæ³¨å†ŒæˆåŠŸ")
                else:
                    logger.warning("âš ï¸ Consulæ³¨å†Œå¤±è´¥ï¼ŒæœåŠ¡ç»§ç»­è¿è¡Œ")

            # 8. å¯åŠ¨gRPCæœåŠ¡å™¨
            server_config = cfg.server
            self.grpc_server = grpc.aio.server(
                futures.ThreadPoolExecutor(max_workers=server_config.max_workers)
            )

            sms_service = SMSService(self.sms_sender)
            sms_pb2_grpc.add_SMSServiceServicer_to_server(sms_service, self.grpc_server)

            self.grpc_server.add_insecure_port(server_config.listen_on)
            await self.grpc_server.start()

            logger.info(f"âœ… gRPCæœåŠ¡å™¨å¯åŠ¨åœ¨ {server_config.listen_on}")
            logger.info(f"ğŸ“± æœåŠ¡åç§°: {server_config.name}")
            logger.info(f"ğŸ”§ è¿è¡Œæ¨¡å¼: {server_config.mode}")

            if cfg.consul.host and cfg.consul.host != "127.0.0.1:8500":
                logger.info(f"ğŸŒ Consulåœ°å€: {cfg.consul.host}")
                logger.info(f"ğŸ—‚ï¸ KVè·¯å¾„: echo_wing/{cfg.server.name}")

            return True

        except Exception as e:
            logger.error(f"âŒ æœåŠ¡å¯åŠ¨å¤±è´¥: {e}")
            return False

    async def run(self):
        """è¿è¡ŒæœåŠ¡ä¸»å¾ªç¯"""
        self._main_task = asyncio.current_task()
        try:
            await self.grpc_server.wait_for_termination()
        except asyncio.CancelledError:
            logger.info("æœåŠ¡ä»»åŠ¡è¢«å–æ¶ˆ")
            raise
        except Exception as e:
            logger.error(f"gRPCæœåŠ¡å™¨å¼‚å¸¸: {e}")
        finally:
            self._main_task = None

    async def stop(self):
        """åœæ­¢å¾®æœåŠ¡"""
        if self._shutting_down:
            return

        self._shutting_down = True
        logger.info("ğŸ›‘ åœæ­¢SMSå¾®æœåŠ¡...")

        # å–æ¶ˆä¸»ä»»åŠ¡
        if self._main_task:
            self._main_task.cancel()
            try:
                await self._main_task
            except asyncio.CancelledError:
                pass

        # æ³¨é”€ConsulæœåŠ¡
        if self.consul_client:
            try:
                await self.consul_client.deregister_service()
                logger.info("âœ… ConsulæœåŠ¡å·²æ³¨é”€")
            except Exception as e:
                logger.error(f"âŒ Consulæ³¨é”€å¤±è´¥: {e}")

        # åœæ­¢gRPCæœåŠ¡å™¨
        if self.grpc_server:
            try:
                # ç«‹å³åœæ­¢ï¼Œä¸å†ç­‰å¾…
                await self.grpc_server.stop(grace=0)
                logger.info("âœ… gRPCæœåŠ¡å™¨å·²åœæ­¢")
            except Exception as e:
                logger.error(f"âŒ åœæ­¢gRPCæœåŠ¡å™¨å¤±è´¥: {e}")

        # æ–­å¼€è°ƒåˆ¶è§£è°ƒå™¨è¿æ¥
        if self.sms_sender:
            try:
                await self.sms_sender.disconnect()
                logger.info("âœ… è°ƒåˆ¶è§£è°ƒå™¨è¿æ¥å·²æ–­å¼€")
            except Exception as e:
                logger.error(f"âŒ æ–­å¼€è°ƒåˆ¶è§£è°ƒå™¨è¿æ¥å¤±è´¥: {e}")

        logger.info("ğŸ‘‹ SMSå¾®æœåŠ¡å·²åœæ­¢")

    def request_shutdown(self):
        """è¯·æ±‚å…³é—­æœåŠ¡"""
        if not self._shutdown_event.is_set():
            self._shutdown_event.set()
            # ç«‹å³å–æ¶ˆä¸»ä»»åŠ¡
            if self._main_task:
                self._main_task.cancel()

    async def wait_for_shutdown(self):
        """ç­‰å¾…å…³é—­ä¿¡å·"""
        try:
            await self._shutdown_event.wait()
        except asyncio.CancelledError:
            pass

    async def _setup_logging(self, log_config):
        """é…ç½®æ—¥å¿—"""
        logger.remove()

        if log_config.mode in ["console", "both"]:
            logger.add(
                sys.stdout,
                format="<green>{time:YYYY-MM-DD HH:mm:ss}</green> | "
                       "<level>{level: <8}</level> | "
                       "<cyan>{name}</cyan>:<cyan>{function}</cyan>:<cyan>{line}</cyan> - "
                       "<level>{message}</level>",
                level=log_config.level.upper(),
                colorize=True
            )

        if log_config.mode in ["file", "both"] and log_config.file_path:
            log_file = Path(log_config.file_path)
            log_file.parent.mkdir(parents=True, exist_ok=True)

            logger.add(
                str(log_file),
                format="{time:YYYY-MM-DD HH:mm:ss} | {level: <8} | "
                       "{name}:{function}:{line} - {message}",
                level=log_config.level.upper(),
                rotation="1 day",
                retention="7 days",
                encoding=log_config.encoding
            )

    async def _print_config(self, cfg):
        """æ‰“å°é…ç½®ä¿¡æ¯"""
        logger.info("\n" + "="*50)
        logger.info("ğŸ“‹ æœåŠ¡é…ç½®:")
        logger.info(f"   æœåŠ¡åç§°: {cfg.server.name}")
        logger.info(f"   ç›‘å¬åœ°å€: {cfg.server.listen_on}")
        logger.info(f"   è¿è¡Œæ¨¡å¼: {cfg.server.mode}")

        if cfg.consul.host and cfg.consul.host != "127.0.0.1:8500":
            logger.info(f"   Consulåœ°å€: {cfg.consul.host}")

        logger.info(f"   ä¸²å£æ³¢ç‰¹ç‡: {cfg.serial.baudrate}")
        logger.info(f"   æ—¥å¿—çº§åˆ«: {cfg.log.level}")
        logger.info("="*50 + "\n")


async def shutdown_handler(service: SMSMicroservice, signum):
    """å¼‚æ­¥ä¿¡å·å¤„ç†å‡½æ•°"""
    logger.info(f"ğŸ“¶ æ”¶åˆ°ä¿¡å· {signum}ï¼Œæ­£åœ¨å…³é—­...")
    service.request_shutdown()


def main():
    """ä¸»å‡½æ•°"""
    import argparse

    parser = argparse.ArgumentParser(description="SMSå¾®æœåŠ¡")
    parser.add_argument("--config", "-c", default="config/sms.yaml",
                       help="é…ç½®æ–‡ä»¶è·¯å¾„")
    args = parser.parse_args()

    # åˆ›å»ºå¾®æœåŠ¡å®ä¾‹
    service = SMSMicroservice(args.config)

    # åˆ›å»ºäº‹ä»¶å¾ªç¯
    loop = asyncio.new_event_loop()
    asyncio.set_event_loop(loop)

    # è®¾ç½®ä¿¡å·å¤„ç†
    for sig in (signal.SIGINT, signal.SIGTERM):
        loop.add_signal_handler(
            sig,
            lambda s=sig: asyncio.create_task(shutdown_handler(service, s))
        )

    async def main_async():
        """å¼‚æ­¥ä¸»å‡½æ•°"""
        # å¯åŠ¨æœåŠ¡
        started = await service.start()
        if not started:
            logger.error("âŒ æœåŠ¡å¯åŠ¨å¤±è´¥")
            sys.exit(1)

        try:
            # åˆ›å»ºå¹¶è¿è¡ŒæœåŠ¡ä»»åŠ¡
            service_task = asyncio.create_task(service.run())

            # ç­‰å¾…å…³é—­ä¿¡å·
            await service.wait_for_shutdown()

            # åœæ­¢æœåŠ¡
            await service.stop()

            # ç­‰å¾…æœåŠ¡ä»»åŠ¡å®Œæˆ
            await service_task
        except asyncio.CancelledError:
            logger.info("ä¸»ä»»åŠ¡è¢«å–æ¶ˆ")
        except Exception as e:
            logger.error(f"ä¸»ç¨‹åºå¼‚å¸¸: {e}")
        finally:
            # æ¸…ç†ä¿¡å·å¤„ç†å™¨
            for sig in (signal.SIGINT, signal.SIGTERM):
                loop.remove_signal_handler(sig)

            logger.info("ğŸ æœåŠ¡å…³é—­å®Œæˆ")

    try:
        # è¿è¡Œä¸»å¾ªç¯
        loop.run_until_complete(main_async())
    except KeyboardInterrupt:
        logger.info("ç¨‹åºè¢«ç”¨æˆ·ä¸­æ–­")
    except Exception as e:
        logger.error(f"ç¨‹åºå¼‚å¸¸: {e}")
        sys.exit(1)
    finally:
        # å…³é—­äº‹ä»¶å¾ªç¯
        loop.close()


if __name__ == "__main__":
    main()
